qcodes/utils/metadata.py

```diff
@@ -42,3 +52,68 @@ def snapshot_base(self, update: bool=False,
         override this with the primary information for a subclass
         """
         return {}
+
+
+class ParameterDiff(NamedTuple):
+    # Cannot be generic in Python < 3.7:
+    # https://stackoverflow.com/questions/50530959/generic-namedtuple-in-python-3-6
+    left_only : ParameterDict[Any]
+    right_only : ParameterDict[Any]
+    changed : ParameterDict[Tuple[Any, Any]]
+
+## FUNCTIONS ##
+
+def extract_param_values(snapshot: Dict[str, Any]) -> Dict[ParameterKey, Any]:
+    """
+    Given a snapshot, returns a dictionary from
+    instrument and parameter names onto parameter values.
+    """
+    parameters = {}
+    for param_name, param in snapshot['station']['parameters'].items():
+        parameters[param_name] = param['value']
+    for instrument_name, instrument in snapshot['station']['instruments'].items():
+        for param_name, param in instrument['parameters'].items():
+            if 'value' in param:
+                parameters[instrument_name, param_name] = param['value']
+
+    return parameters
+
+def diff_param_values(left_snapshot: Dict[str, Any],
```

Comment:
Should the type hint be used here?
```suggestion
def diff_param_values(left_snapshot: Snapshot[Any],
```